File: geode-client-native/quickstart/cpp/queryobjects/PortfolioPdx.cpp
#include "PortfolioPdx.hpp"
#include <malloc.h>

using namespace gemfire;
using namespace testobject;

const char*  PortfolioPdx::secIds[] = { "SUN", "IBM", "YHOO", "GOOG", "MSFT",
      "AOL", "APPL", "ORCL", "SAP", "DELL"};

PortfolioPdx::PortfolioPdx(int32_t i, int32_t size, char** nm) : names(nm) {
    id = i;

    char pkidbuf[256];
    sprintf(pkidbuf,"%d",i);
    size_t strSize = strlen(pkidbuf) + 1;
    pkid = new char[strSize];
    memcpy(pkid, pkidbuf, strSize);

    const char* statusStr = (i % 2 == 0) ? "active" : "inactive";
    int32_t statusSize = static_cast<int32_t> (strlen( statusStr )) + 1;
    status = new char[ statusSize ];
    memcpy( status, statusStr, statusSize );

    char buf[100];
    sprintf( buf, "type%d", (i % 3) );
    size_t strSize2 = strlen(buf) + 1;
    type = new char[strSize2];
    memcpy(type, buf, strSize2);

    int numSecIds = sizeof(secIds)/sizeof(char*);
    position1 = new PositionPdx(secIds[PositionPdx::cnt % numSecIds], PositionPdx::cnt * 1000);
    if (i % 2 != 0) {
      position2 = new PositionPdx(secIds[PositionPdx::cnt % numSecIds],PositionPdx::cnt * 1000);
    }
    else
    {
      position2 = NULLPTR;
    }
    positions = CacheableHashMap::create( );
    positions->insert(CacheableString::create(secIds[PositionPdx::cnt % numSecIds]), position1);

    if(size > 0){
      newVal = new int8_t[size];
      for(int index =0; index < size; index++){
        newVal[index] = (int8_t)'B';
      }
    }
    newValSize = size;

    time_t timeVal = 1310447869;
    creationDate = CacheableDate::create( timeVal  );
    arrayNull = NULL;
    arrayZeroSize = new int8_t[0];

}

PortfolioPdx::~PortfolioPdx() {
  if ( newVal != NULL ) {
    delete [] newVal;
    newVal = NULL;
  }
  if ( status != NULL ) {
    delete [] status;
    status = NULL;
  }
  if ( pkid != NULL ) {
      delete [] pkid;
      pkid = NULL;
  }

  if ( type != NULL ) {
    delete [] type;
    type = NULL;
  }

  if ( newVal != NULL ) {
    delete [] newVal;
    newVal = NULL;
  }
}

void PortfolioPdx::toData( PdxWriterPtr pw  )  {
  pw->writeInt("ID", id);
  pw->markIdentityField("ID");

  pw->writeString("pkid", pkid);
  pw->markIdentityField("pkid");

  pw->writeObject("position1", position1);
  pw->markIdentityField("position1");

  pw->writeObject("position2", position2);
  pw->markIdentityField("position2");

  pw->writeObject("positions", positions);
  pw->markIdentityField("positions");

  pw->writeString("type", type);
  pw->markIdentityField("type");

  pw->writeString("status", status);
  pw->markIdentityField("status");

  pw->writeStringArray("names", names, 0);
  pw->markIdentityField("names");

  pw->writeByteArray("newVal", newVal, newValSize);
  pw->markIdentityField("newVal");

  pw->writeDate("creationDate", creationDate);
  pw->markIdentityField("creationDate");

  pw->writeByteArray("arrayNull", arrayNull,0);
  pw->writeByteArray("arrayZeroSize", arrayZeroSize,0);

}

void PortfolioPdx::fromData( PdxReaderPtr pr )
{
  id = pr->readInt("ID");
  pkid = pr->readString("pkid");

  position1 = dynCast<PositionPdxPtr>(pr->readObject("position1"));
  position2 = dynCast<PositionPdxPtr>(pr->readObject("position2"));
  positions = dynCast<CacheableHashMapPtr>(pr->readObject("positions"));
  type = pr->readString("type");
  status = pr->readString("status");

  int32_t strLenArray=0;
  names = pr->readStringArray("names", strLenArray);
  int32_t byteArrayLen=0;
  newVal = pr->readByteArray("newVal", byteArrayLen);
  creationDate = pr->readDate("creationDate");
  int32_t arrayNullLen = 0;
  arrayNull = pr->readByteArray("arrayNull", arrayNullLen);
  int32_t arrayZeroSizeLen=0;
  arrayZeroSize = pr->readByteArray("arrayZeroSize", arrayZeroSizeLen);

}

CacheableStringPtr PortfolioPdx::toString() const {
  LOGINFO("PortfolioPdx::toString() Start");
  char idbuf[1024];
  sprintf(idbuf,"PortfolioPdxObject: [ id=%d ]",id);

  char pkidbuf[1024];
  if (pkid != NULL) {
    sprintf(pkidbuf, " status=%s type=%s pkid=%s\n", this->status,
        this->type, this->pkid);
  }
  else {
    sprintf(pkidbuf, " status=%s type=%s pkid=%s\n", this->status,
        this->type, this->pkid);
  }
  char position1buf[2048];
  if (position1 != NULLPTR) {
    sprintf(position1buf, "\t\t\t  P1: %s", position1->toString()->asChar());
  }
  else {
    sprintf(position1buf, "\t\t\t  P1: %s", "NULL");
  }
  char position2buf[2048];
  if (position2 != NULLPTR) {
    sprintf(position2buf, " P2: %s", position2->toString()->asChar());
  }
  else {
    sprintf(position2buf, " P2: %s ]", "NULL");
  }
  char creationdatebuf[2048];
  if (creationDate != NULLPTR) {
    sprintf(creationdatebuf, "creation Date %s",
        creationDate->toString()->asChar());
  }
  else {
    sprintf(creationdatebuf, "creation Date %s", "NULL");
  }

  char stringBuf[7000];
  sprintf(stringBuf, "%s%s%s%s%s",idbuf,pkidbuf,creationdatebuf,position1buf,position2buf);
  return CacheableString::create( stringBuf );

  return CacheableString::create( idbuf );
}

File: geode-client-native/src/tests/clicache/PdxClassLibrary/PortfolioPdx.cs
//=========================================================================
// Copyright (c) 2002-2014 Pivotal Software, Inc. All Rights Reserved.
// This product is protected by U.S. and international copyright
// and intellectual property laws. Pivotal products are covered by
// more patents listed at http://www.pivotal.io/patents.
//========================================================================

using System;
using System.Collections;
using GemStone.GemFire.Cache.Generic;


namespace PdxTests
{
  /// <summary>
  /// User class for testing the put functionality for object.
  /// </summary>
  public class PortfolioPdx
    : IPdxSerializable
  {
    #region Private members and methods

    private int m_id;
    private string m_pkid;
    private PositionPdx m_position1;
    private PositionPdx m_position2;
    private Hashtable m_positions;
    private string m_type;
    private string m_status;
    private string[] m_names;
    private byte[] m_newVal;
    private DateTime m_creationDate;
    private byte[] m_arrayZeroSize;
    private byte[] m_arrayNull;

    private static string[] m_secIds = { "SUN", "IBM", "YHOO", "GOOG", "MSFT",
      "AOL", "APPL", "ORCL", "SAP", "DELL" };

    private UInt32 GetObjectSize(IGFSerializable obj)
    {
      return (obj == null ? 0 : obj.ObjectSize);
    }

    #endregion

    #region Public accessors

    public int ID
    {
      get
      {
        return m_id;
      }
    }

    public string Pkid
    {
      get
      {
        return m_pkid;
      }
    }

    public PositionPdx P1
    {
      get
      {
        return m_position1;
      }
    }

    public PositionPdx P2
    {
      get
      {
        return m_position2;
      }
    }

    public Hashtable Positions
    {
      get
      {
        return m_positions;
      }
    }

    public string Status
    {
      get
      {
        return m_status;
      }
    }

    public bool IsActive
    {
      get
      {
        return (m_status == "active");
      }
    }

    public byte[] NewVal
    {
      get
      {
        return m_newVal;
      }
    }

    public byte[] ArrayNull
    {
      get
      {
        return m_arrayNull;
      }
    }

    public byte[] ArrayZeroSize
    {
      get
      {
        return m_arrayZeroSize;
      }
    }

    public DateTime CreationDate
    {
      get
      {
        return m_creationDate;
      }
    }

    public string Type
    {
      get
      {
        return m_type;
      }
    }

    public static string[] SecIds
    {
      get
      {
        return m_secIds;
      }
    }

    public override string ToString()
    {
      string portStr = string.Format("Portfolio [ID={0} status={1} " +
        "type={2} pkid={3}]", m_id, m_status, m_type, m_pkid);
      portStr += string.Format("{0}\tP1: {1}",
        Environment.NewLine, m_position1);
      portStr += string.Format("{0}\tP2: {1}",
        Environment.NewLine, m_position2);
      portStr += string.Format("{0}Creation Date: {1}", Environment.NewLine,
        m_creationDate);
      return portStr;
    }

    #endregion

    #region Constructors

    public PortfolioPdx()
    {
      m_id = 0;
      m_pkid = null;
      m_type = null;
      m_status = null;
      m_newVal = null;
      m_creationDate = new DateTime();
    }

    public PortfolioPdx(int id)
      : this(id, 0)
    {
    }

    public PortfolioPdx(int id, int size)
      : this(id, size, null)
    {
    }

    public PortfolioPdx(int id, int size, string[] names)
    {
      m_names = names;
      m_id = id;
      m_pkid = id.ToString();
      m_status = (id % 2 == 0) ? "active" : "inactive";
      m_type = "type" + (id % 3);
      int numSecIds = m_secIds.Length;
      m_position1 = new PositionPdx(m_secIds[PositionPdx.Count % numSecIds],
        PositionPdx.Count * 1000);
      if (id % 2 != 0)
      {
        m_position2 = new PositionPdx(m_secIds[PositionPdx.Count % numSecIds],
          PositionPdx.Count * 1000);
      }
      else
      {
        m_position2 = null;
      }
      m_positions = new Hashtable();
      m_positions[m_secIds[PositionPdx.Count % numSecIds]] =
        m_position1;
      if (size > 0)
      {
        m_newVal = new byte[size];
        for (int index = 0; index < size; index++)
        {
          m_newVal[index] = (byte)'B';
        }
      }
      m_creationDate = DateTime.Now;
      m_arrayNull = null;
      m_arrayZeroSize = new byte[0];
    }

    #endregion
      
    public static IPdxSerializable CreateDeserializable()
    {
      return new PortfolioPdx();
    }

    #region IPdxSerializable Members

    public void FromData(IPdxReader reader)
    {
      m_id = reader.ReadInt("id");

      bool isIdentity =  reader.IsIdentityField("id");

      if (isIdentity == false)
        throw new IllegalStateException("Portfolio id is identity field");

      bool isId = reader.HasField("id");

      if (isId == false)
        throw new IllegalStateException("Portfolio id field not found");

      bool isNotId = reader.HasField("ID");

      if (isNotId == true)
        throw new IllegalStateException("Portfolio isNotId field found");

      m_pkid = reader.ReadString("pkid");
      m_position1 = (PositionPdx)reader.ReadObject("position1");
      m_position2 = (PositionPdx)reader.ReadObject("position2");
      m_positions = (Hashtable)reader.ReadObject("positions");
      m_type = reader.ReadString("type");
      m_status = reader.ReadString("status");
      m_names = reader.ReadStringArray("names");
      m_newVal = reader.ReadByteArray("newVal");
      m_creationDate = reader.ReadDate("creationDate");
      m_arrayNull = reader.ReadByteArray("arrayNull");
      m_arrayZeroSize = reader.ReadByteArray("arrayZeroSize");
    }

    public void ToData(IPdxWriter writer)
    {
      writer.WriteInt("id", m_id)
      //identity field
      .MarkIdentityField("id")
      .WriteString("pkid", m_pkid)
      .WriteObject("position1", m_position1)
      .WriteObject("position2", m_position2)
      .WriteObject("positions", m_positions)
      .WriteString("type", m_type)
      .WriteString("status", m_status)
      .WriteStringArray("names", m_names)
      .WriteByteArray("newVal", m_newVal)
      .WriteDate("creationDate", m_creationDate)
      .WriteByteArray("arrayNull", m_arrayNull)
      .WriteByteArray("arrayZeroSize", m_arrayZeroSize);      
    }

    #endregion
  }
}
